File: src/java/10026.java
// 백준 10026번 문제 : 적록색약
// https://www.acmicpc.net/problem/10026
// 알고리즘 분류 : 그래프 이론, 그래프 탐색, 너비 우선 탐색, 깊이 우선 탐색

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Pair // Implement pair in C++
{
    private int y;
    private int x;
    
    public Pair(int init_y, int init_x)
    {
        y = init_y;
        x = init_x;
    }
    
    public int first()
    {
        return y;
    }
    
    public int second()
    {
        return x;
    }
}

public class Main
{
    private static char[][] colorMap;
    private static boolean[][][] isVisited;
    private static int n;
    private static int[] yMove = {-1, 0, 1, 0};
    private static int[] xMove = {0, 1, 0, -1};
    
    public static int getArea(int isColorBlind)
    {
        int count = 0;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (!isVisited[i][j][isColorBlind]) // BFS based task.
                {
                    char initColor = colorMap[i][j];
                    if (isColorBlind == 1 && colorMap[i][j] == 'R') initColor = 'G';
                    
                    Queue<Pair> q = new LinkedList<Pair>();
                    
                    isVisited[i][j][isColorBlind] = true;
                    q.offer(new Pair(i, j));
                    while (!q.isEmpty())
                    {
                        Pair p = q.poll();
                        int y = p.first();
                        int x = p.second();
                        
                        for (int k = 0; k < 4; k++)
                        {
                            int n_y = y + yMove[k];
                            int n_x = x + xMove[k];
                            if (n_y < 0 || n_y >= n || n_x >= n || n_x < 0 || isVisited[n_y][n_x][isColorBlind] == true)
                                continue;
                            
                            if (isColorBlind == 1 && (initColor == 'G' && colorMap[n_y][n_x] != 'B' || initColor == 'B' && colorMap[n_y][n_x] == initColor))
                            {
                                isVisited[n_y][n_x][isColorBlind] = true;
                                q.offer(new Pair(n_y, n_x));
                            }
                            if (isColorBlind == 0 && initColor == colorMap[n_y][n_x])
                            {
                                isVisited[n_y][n_x][isColorBlind] = true;
                                q.offer(new Pair(n_y, n_x));
                            }
                        }
                    }
                    count++;
                }
            }
        }
        return count;
    }
    
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        n = Integer.parseInt(br.readLine());
        colorMap = new char[n][];
        isVisited = new boolean[n][n][2];
        for (int i = 0; i < n; i++)
        {
            String input = br.readLine();
            colorMap[i] = input.toCharArray();
        }
        
        System.out.printf("%d %d\n", getArea(0), getArea(1));
    }
}

File: src/csharp/10026.cs
// 백준 10026번 문제 : 적록색약
// https://www.acmicpc.net/problem/10026
// 알고리즘 분류 : 그래프 이론, 그래프 탐색, 너비 우선 탐색, 깊이 우선 탐색

var yMove = new[] { 0, 1, 0, -1 };
var xMove = new[] { 1, 0, -1, 0 };
int zoneCount = 0, colorBlindZoneCount = 0;

int n = int.Parse(Console.ReadLine());
var board = new string[n];
var isVisited = new bool[n, n, 2];

for (int i = 0; i < n; i++)
    board[i] = Console.ReadLine();
for (int i = 0; i < n; i++)
{
    for (int j = 0; j < n; j++)
    {
        if (!isVisited[i, j, 0])
        {
            BFS(i, j);
            zoneCount++;
        }
        if (!isVisited[i, j, 1])
        {
            BFS(i, j, 1);
            colorBlindZoneCount++;
        }
    }
}
Console.WriteLine($"{zoneCount} {colorBlindZoneCount}");

void BFS(int y, int x, int isColorBlind = 0)
{
    var q = new Queue<(int y, int x)>();
    isVisited[y, x, isColorBlind] = true;
    q.Enqueue((y, x));
    char initColor = (isColorBlind == 1 && board[y][x] != 'B') ? 'A' : board[y][x];

    while (q.Count > 0)
    {
        (y, x) = q.Dequeue();
        for (int i = 0; i < 4; i++)
        {
            int newY = y + yMove[i];
            int newX = x + xMove[i];

            if (newX < 0 || newY < 0 || newX >= n || newY >= n) continue;
            if (!isVisited[newY, newX, isColorBlind] && ((initColor == 'A' && board[newY][newX] != 'B') || board[newY][newX] == initColor))
            {
                isVisited[newY, newX, isColorBlind] = true;
                q.Enqueue((newY, newX));
            }
        }
    }
}
